GVFS/GVFS.Virtualization/FileSystemCallbacks.cs

```diff
@@ -256,36 +260,28 @@ public bool IsReadyForExternalAcquireLockRequests(NamedPipeMessages.LockData req
             return true;
         }
 
-        public EventMetadata GetMetadataForHeartBeat(ref EventLevel eventLevel)
+        public EventMetadata GetAndResetHeartBeatMetadata(out bool logToFile)
         {
+            logToFile = false;
             EventMetadata metadata = new EventMetadata();
-            if (this.placeHolderCreationCount.Count > 0)
-            {
-                ConcurrentDictionary<string, PlaceHolderCreateCounter> collectedData = this.placeHolderCreationCount;
-                this.placeHolderCreationCount = new ConcurrentDictionary<string, PlaceHolderCreateCounter>(StringComparer.OrdinalIgnoreCase);
-
-                int count = 0;
-                foreach (KeyValuePair<string, PlaceHolderCreateCounter> processCount in
-                    collectedData.OrderByDescending((KeyValuePair<string, PlaceHolderCreateCounter> kvp) => kvp.Value.Count))
-                {
-                    ++count;
-                    if (count > 10)
-                    {
-                        break;
-                    }
-
-                    metadata.Add("ProcessName" + count, processCount.Key);
-                    metadata.Add("ProcessCount" + count, processCount.Value.Count);
-                }
 
-                eventLevel = EventLevel.Informational;
-            }
+            metadata.Add(
+                "FilePlaceholderCreation",
+                this.GetProcessInteractionData(this.GetAndResetProcessCountMetadata(ref this.filePlaceHolderCreationCount), ref logToFile));

```

Comment:
Another thought here.

Do we really need to clear the list between each heartbeat?  Would it be more beneficial to keep a running total of largest hydrators across the mount?  You can get all the same information with either approch but a running counter would be simplier.  Both approaches would have pros/cons.   @wilbaker

Reply:
The historical reasons for clearing each heartbeat:

- We only want to write the heartbeat to the log file when something has changed, and starting from scratch for each heartbeat makes it easier to know if anything has changed in subsequent heartbeats.

- When helping out users you'll be able to immediately tell which processes created placeholders in each time slice (without having to search for older heartbeats and calculate the new values).

- Our main customer was processing this data as well, and resetting it each heartbeat was helpful for the diagnostics/reporting.

My personal preference is to continue to reset each heartbeat, because I've found it easier to reason about what happened when I look at the log files.